File: src/main/java/org/graylog2/indexer/searches/Searches.java
/**
 * Copyright 2013 Lennart Koopmann <lennart@socketfeed.com>
 *
 * This file is part of Graylog2.
 *
 * Graylog2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog2.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.graylog2.indexer.searches;

import org.elasticsearch.action.search.SearchRequestBuilder;
import org.elasticsearch.action.search.SearchResponse;
import org.elasticsearch.client.Client;
import org.graylog2.Core;

import static org.elasticsearch.index.query.QueryBuilders.queryString;

/**
 * @author Lennart Koopmann <lennart@socketfeed.com>
 */
public class Searches {

	private final Core server;
	private final Client c;
	
	public Searches(Client client, Core server) {
		this.server = server;
		this.c = client;
	}
	
	public SearchResult universalSearch(String query) {
		SearchRequestBuilder srb = c.prepareSearch();
		srb.setIndices(server.getDeflector().getAllDeflectorIndexNames()); // XXX 020: have a method that builds time ranged index requests
		srb.setQuery(queryString(query));

		SearchResponse r = c.search(srb.request()).actionGet();
		return new SearchResult(r.hits(), query, r.took());
	}
	
}

File: graylog2-web-interface/test/fixtures/searches.ts
/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */

/* eslint-disable import/prefer-default-export */

import { Map } from 'immutable';

import Query from 'views/logic/queries/Query';
import Search from 'views/logic/search/Search';
import View from 'views/logic/views/View';
import ViewState from 'views/logic/views/ViewState';
import type _Widget from 'views/logic/widgets/Widget';
import type { WidgetPositions } from 'views/types';

export const createSearch = ({ searchId, queryId }: { searchId?: string, queryId?: string} = {}) => {
  const exampleSearchId = searchId ?? 'search-id-1';
  const exampleQueryId = queryId ?? 'query-id-1';

  const viewState = ViewState.builder()
    .titles(Map())
    .build();
  const query = Query.builder().id(exampleQueryId).build();
  const searchSearch = Search.builder().queries([query]).id(exampleSearchId).build();

  return View.builder()
    .search(searchSearch)
    .type(View.Type.Dashboard)
    .state(Map({ [exampleQueryId]: viewState }))
    .id(exampleSearchId)
    .title('Search 1')
    .build();
};

export const createViewWithWidgets = (widgets: Array<_Widget>, positions: WidgetPositions) => {
  const view = createSearch();
  const newViewState = view.state.get('query-id-1')
    .toBuilder()
    .widgets(widgets)
    .widgetPositions(positions)
    .build();

  return view
    .toBuilder()
    .state(Map({ 'query-id-1': newViewState }))
    .build();
};
